### PR TITLE
[Tabs] Allow overriding internal size class.

### PR DESCRIPTION
**Pull request by romoore**
The internals of MDCTabBar depends on the view's trait collection's
horizontal size class to determine its leading/trailing insets for some
layout types. As a means of unblocking some internal customizations,
this change introduces a new delegate that allows overriding the
horizontal size class.

Clients should not use these APIs without receiving explicit approval
from the MDC-iOS team. These APIs are likely to be removed in a future
release, though without a specific timeline.

This is proposed as an alternative approach to https://github.com/material-components/material-components-ios/pull/7515, which uses subclassing explicitly.

### CODE REVIEW

File: components/Tabs/src/private/MDCItemBar.h
Comment (mrhappyasthma) [inline components/Tabs/src/private/MDCItemBar.h]: The linewrapping here seems off for the comment.

File: components/Tabs/src/private/MDCItemBar.m
Comment (mrhappyasthma) [inline components/Tabs/src/private/MDCItemBar.m]: The check to `self.tabBar` is not needed since if it's nil the sizeClassDelegate check will be nil and fail automatically. So it's implicitly checked by the later arguments.

This could be simplified to:

```
if ([self.tabBar.sizeClassDelegate respondsToSelector:@selector(horizontalSizeClassForObject:)] {

}
```

And it would have the same effect.

Comment (mrhappyasthma) [inline components/Tabs/src/private/MDCItemBar.m]: I'd consider using a local variable to hold a reference to the delegate to avoid TOCTOU issues. I also think it might simplify readability: http://go/gh/stephanemoore/archives/blob/master/objc/pitfalls/tocttou.md

```
id<MDCTabBarSizeClassDelegate> sizeClassDelegate = self.tabBar.sizeClassDelegate;
if ([sizeClassDelegate respondsToSelector:@selector(horizontalSizeClassForObject:)]) {
  return [sizeClassDelegate horizontalSizeClassForObject:self];
}
```

**Review by mrhappyasthma — APPROVED** [review]
Generally looks good to me. My original PR was more conservative and didn't want to update the headers since I don't have the context on the long-term goal of these classes, so I tried to avoid as much maintenance overhead as possible.

I like this alternative better since it's a more explicit contract between the delegates. If your team is okay with exposing this delegate and maintaining it in the short-term, then this works great for me.

File: components/Tabs/src/private/MDCItemBar.m
Comment (romoore) [reply components/Tabs/src/private/MDCItemBar.m]: Thanks, that seems nearly impossible for single-thread code like this (the linked doc didn't explain how single-threading could have these race conditions),  I'll update with an automatic variable.

**Comment by romoore** [comment]
@mrhappyasthma I'm doing my best to bury the new APIs in a header that isn't included by the umbrella. That should allow us sufficient visibility restrictions using Bazel visibility rules. The header docs I think also make it clear that clients shouldn't depend on this without review by the maintainers.  I do believe we generally prefer more explicit contracts like this and to avoid _required_ subclassing whenever we can.

File: components/Tabs/src/private/MDCItemBar.h
Comment (romoore) [reply components/Tabs/src/private/MDCItemBar.h]: Done, good catch!

File: components/Tabs/src/private/MDCItemBar.m
Comment (romoore) [reply components/Tabs/src/private/MDCItemBar.m]: Done, thanks!

**Comment by romoore** [comment]
@mrhappyasthma @jverkoey Do we think this approach (returning size class) is better than, for example, requesting that the delegate return UIEdgeInsets for a specified layout + UITraitEnvironment?

**Comment by mrhappyasthma** [comment]
I think returning edge inset values directly gives the most flexibility, which is probably better than returning just the size class value. (For instance we don't really use size classes at all, but forcing Compact all the time without any other changes to MDCItemBar is suitable for our needs.) I went with the size class approach in my PR because I was trying to come up with the most minimal change to accomplish what we needed.

I imagine you would need a slightly larger refactor of how the insets are calculated in MDCItemBar to accomplish this, so I'd be supportive of that if it's something you're willing to do. But it's not strictly required to fix the immediate bugs in our project, since the insets for the Compact class are desirable in all cases for our app, as it exists today.

**Comment by romoore** [comment]
Thanks, Mark. @jverkoey I'd like to hear your thoughts. 

The current internal structure of MDCTabBar is very likely to change in the near future, so a more flexible API is likely to be written in any case.  It may be sufficient to expose this near-hack for now, then migrate any clients using it to improved APIs later. As you mentioned, Mark, our goal here is to unblock your use cases without unnecessary code changes.

**Review by jverkoey — APPROVED** [review]

**Comment by jverkoey** [comment]
> Do we think this approach (returning size class) is better than, for example, requesting that the delegate return UIEdgeInsets for a specified layout + UITraitEnvironment?

I think the API in this PR is a reasonable starting level of abstraction, and it sounds like it meets @mrhappyasthma's needs albeit not to the degree of flexibility that may support other future needs.

@mrhappyasthma so long as are you able to achieve what you need to with this approach we'll pursue this for now with the intent of keeping an eye out for more flexible improvements as part of our broader body of work to improve Tabs.

**Comment by mrhappyasthma** [comment]
Sounds reasonable to me.

**Pull request merged**
